File: binarytreeverticalordertraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def verticalOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        self.traverse = []
        self.preorder(root, 0, 0)
        #print self.traverse
        self.traverse.sort(key = lambda x: x[1])
        #print self.traverse
        visited = set()
        
        offset = self.traverse[0][1] - 0
        
        for i in self.traverse:
            i[1] -= offset
        res = []
        for i in self.traverse:
            if i[1] == len(res):
                res.append([])
            res[i[1]].append((i[0], i[2]))
        print res
        for i in xrange(len(res)):
            res[i].sort(key = lambda x: x[1], reverse = True)
            res[i] = [j[0] for j in res[i]]
        #print res
        return res

    def preorder(self, root, level, height):
        if not root:
            return
        
        self.preorder(root.left, level-1, height-1)
        self.traverse.append([root.val, level, height])
        self.preorder(root.right, level+1, height-1)

File: binarytreeverticalordertraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        map<int, vector<int>> vt;
        vector<vector<int>> res;
        if (!root) return res;
        queue<pair<int, TreeNode*>> que;
        que.push(make_pair(0, root));
        while (!que.empty()) {
            int n = que.size();
            for (int i = 0; i < n; i++) {
                auto target = que.front();
                que.pop();
                vt[target.first].push_back(target.second->val);
                if (target.second->left) que.push(make_pair(target.first-1, target.second->left));
                if (target.second->right) que.push(make_pair(target.first+1, target.second->right));
            }
            
        }
        for (auto it = vt.begin(); it != vt.end(); it++) res.push_back(it->second);
        return res;
    }
};
